library/std/src/sys/solid/fs.rs

```diff
@@ -339,6 +339,29 @@ impl File {
         }
     }
 
+    pub fn read_buf(&self, buf: &mut ReadBuf<'_>) -> io::Result<()> {
+        let num_bytes_read = unsafe {
+            let len = buf.remaining();
+            let mut out_num_bytes = MaybeUninit::uninit();
+            error::SolidError::err_if_negative(abi::SOLID_FS_Read(
+                self.fd.raw(),
+                buf.unfilled_mut().as_mut_ptr() as *mut u8,
+                len,
+                out_num_bytes.as_mut_ptr(),
+            ))
+            .map_err(|e| e.as_io_error())?;
+            out_num_bytes.assume_init()
+        };
+
+        // Safety: `num_bytes_read` bytes were written to the unfilled
+        // portion of the buffer
+        unsafe { buf.assume_init(num_bytes_read) };
```

Comment:
nit: I would just have a single unsafe block rather than splitting it into two - afaict there is no benefit from the split, and the safety invariant in each block is not independent.

Reply:
Agreed, updated the PR.